Adding STM32L071xx target

andreagilardoni opened the pull request:
Adding STM32L071CXCTX as a generic target to be extended.
This addition required to fix some issues on stml0 library

<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Added a new generic target for STM32L071CXCTX based boards. If you want to use this you have to create a custom target that inherits from this target.

#### Impact of changes <!-- Optional -->

#### Migration actions required <!-- Optional -->

### Documentation <!-- Required -->

The changes performed are on the following files/directories:

- ``targets/TARGET_STM/TARGET_STM32L0/device/stm32l0xx_hal_conf.h``: some HAL features had to be disabled since the MCU is not using them
- ``targets/TARGET_STM/TARGET_STM32L0/device/stm32l0xx_hal_tsc.h``: changed the order of preprocessor directives to for ``extern "C" { ... }`` definition, since if TSC was disabled the curly bracket (`}`) was missing causing all sort of problems
- ``targets/targets.json``: added the target definition
- ``targets/TARGET_STM/TARGET_STM32L0/TARGET_STM32L071CXCTX/``: added the target definition files

I performed the changes on mbed-os-6.2.0, the next step is to adapt the target definition for mbed-os-6.2.1, where STM related target definitions have been improved following the various STM32 mcu families.

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [X] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR

I still have not received yet the board using the proper MCU, I performed the unit test with the following command: 
```
mbed test -m TEST -t GCC_ARM -v --test-config mbed-os/TESTS/configs/baremetal.json
```

I was not able to test against all the available compilers, since I only have the GCC_ARM, but I followed the STM  [guidelines](https://os.mbed.com/teams/ST/wiki/steps-to-create-a-new-STM32-platform). 

The test were performed by creating a dummy custom test target. The required files are here attached: [tests.zip](https://github.com/ARMmbed/mbed-os/files/5142590/tests.zip), [test.tar.gz](https://github.com/ARMmbed/mbed-os/files/5142601/test.tar.gz).

The test result log are the following: [test_results.log](https://github.com/ARMmbed/mbed-os/files/5142607/test_results.log)


----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32L0/TARGET_STM32L071CXCTX/device/TOOLCHAIN_ARM/startup_stm32l073xx.S:
```diff
@@ -0,0 +1,218 @@
+;******************** (C) COPYRIGHT 2016 STMicroelectronics ********************
+;* File Name          : startup_stm32l073xx.s
```
Copy file from 
https://github.com/STMicroelectronics/STM32CubeL0/blob/master/Drivers/CMSIS/Device/ST/STM32L0xx/Source/Templates/arm/startup_stm32l071xx.s

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32L0/TARGET_STM32L071CXCTX/device/TOOLCHAIN_ARM/stm32l073xz.sct:
```diff
@@ -0,0 +1,53 @@
+#! armcc -E
```
Rename file to L071

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32L0/TARGET_STM32L071CXCTX/device/TOOLCHAIN_IAR/startup_stm32l073xx.S:
```diff
@@ -0,0 +1,348 @@
+;/******************** (C) COPYRIGHT 2016 STMicroelectronics ********************
+;* File Name          : startup_stm32l073xx.s
```
wrong file

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32L0/TARGET_STM32L071CXCTX/device/TOOLCHAIN_IAR/stm32l073xx.icf:
```diff
@@ -0,0 +1,59 @@
+/* Linker script to configure memory regions.
+ *
```
rename file

jeromecoutant commented on targets/targets.json:
```diff
@@ -2736,7 +2736,7 @@
         "mbed_rom_start": "0x08000000",
         "mbed_rom_size": "0x200000",
-        "mbed_ram_start": "0x24000000",
+        "mbed_am_start": "0x24000000",
```
error I think :-)

andreagilardoni replied:
thanks

jeromecoutant commented on targets/targets.json:
```diff
@@ -3033,6 +3033,48 @@
         ],
         "device_name": "STM32L072CZ"
     },
+    "STM32L071CXCTX": {
```
I think, you don't really add a new target?
So please, renam to MCU_STM32L071xx (like MCU_STM32F103xB for ex)
Then in your custom_target.json, you will inherit from it.

andreagilardoni replied:
ok, got it

jeromecoutant commented on targets/targets.json:
```diff
@@ -3033,6 +3033,48 @@
     },
+    "STM32L071CXCTX": {
+        "inherits": [
+            "MCU_STM32"
```
Inherit from MCU_STM32L0,
and then remove duplicated lines

jeromecoutant commented on targets/targets.json:
```diff
@@ -3033,6 +3033,48 @@
+        "mbed_rom_start": "0x8000000",
+        "mbed_rom_size": "0x30000",
+        "mbed_ram_start": "0x20000000",
+        "mbed_ram_size": "0x5000"
```
Should be not useful if device_name is well set

andreagilardoni replied:
ok, I had "mbed-cli" complaining because it didn't have this, does it serve a purpose?

jeromecoutant reviewed (changes requested):
Very good starting point!
Thx

andreagilardoni committed:
Adding a new target

Adding STM32L071CXCTX as a generic target to be extended.
This addition required to fix some issues on stml0 library

andreagilardoni committed:
fixing the issues

andreagilardoni committed:
renaming the mcu to stm32l071xx

andreagilardoni committed:
fixing the remaining parts after the rename

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32L0/device/stm32l0xx_hal_conf.h:
```diff
@@ -53,41 +53,54 @@
 #define HAL_CRYP_MODULE_ENABLED
-#if !defined (TARGET_STM32L031K6) && !defined (TARGET_STM32L011K4)
+
+#if !defined (TARGET_STM32L031K6) && !defined (TARGET_STM32L011K4) && !defined (TARGET_STML071xx)
```
TARGET_STML071xx => TARGET_STM32L071xx

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32L0/device/stm32l0xx_hal_conf.h:
```diff
@@ -53,41 +53,54 @@
+#if !defined (TARGET_STM32L031K6) && !defined (TARGET_STM32L011K4) && !defined (TARGET_STML071xx)
 #define HAL_DAC_MODULE_ENABLED
-#endif /* !TARGET_STM32L031K6  && !TARGET_STM32L011K4 */
+#endif /* !TARGET_STM32L031K6  && !TARGET_STM32L011K4 && !TARGET_STML0CXCTX */
```
TARGET_STM32L071xx

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32L0/device/stm32l0xx_hal_conf.h:
```diff
@@ -53,41 +53,54 @@
 #define HAL_I2C_MODULE_ENABLED
-#if !defined (TARGET_STM32L031K6) && !defined (TARGET_STM32L011K4)
+
+#if !defined (TARGET_STM32L031K6) && !defined (TARGET_STM32L011K4) && !defined (TARGET_STML071xx) 
```
TARGET_STM32L071xx

jeromecoutant commented on targets/targets.json:
```diff
@@ -2786,6 +2786,21 @@
+        "public": false,
+        "extra_labels_add": [
+            "STM32L0",
+            "STML071xx"
```
STM32L071xx

jeromecoutant commented on targets/targets.json:
```diff
@@ -2786,6 +2786,21 @@
+            "ANALOGOUT",
+            "TRNG"
+        ],
+        "device_name": "STM32L071xx"
```
In fact, you have to remove "device_name" from "MCU_STM32L071xx",
as device name is the exact used MCU, and STM32L071xx is the family

This "device_name" has to be defined at target level, so in the custom_target.json file.
This parameter is used for bootloader info, and for debug with pyOCD and MBED Studio for ex.
It has to be part of https://github.com/ARMmbed/mbed-os/blob/master/tools/arm_pack_manager/index.json
So maybe STM32L071CZTx for you ?

andreagilardoni replied:
ok, got it

andreagilardoni committed:
fixing small issues

jeromecoutant reviewed (approved):
For my information, do you an information webpage for your custom target ?
Thx

0xc0170 commented on targets/TARGET_STM/TARGET_STM32L0/TARGET_MCU_STM32L071xx/objects.h:
```diff
@@ -0,0 +1,63 @@
+/* mbed Microcontroller Library
+ *******************************************************************************
+ * Copyright (c) 2015, STMicroelectronics
```
Please add SPDX identifier to new files like this

0xc0170 commented on targets/TARGET_STM/TARGET_STM32L0/TARGET_MCU_STM32L071xx/device/us_ticker_data.h:
```diff
@@ -0,0 +1,45 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2018 ARM Limited
```
Please add SPDX identifier to new files like this

I understand this might be copy-paste from some older file (still needs to be fixed in targets)

andreagilardoni replied:
ok, In order to avoid making any mistakes on this how should it look like?

Like this?
```
  /* mbed Microcontroller Library
   *******************************************************************************
   * Copyright (c) 2015, STMicroelectronics
   * All rights reserved.
   *
   * Redistribution and use in source and binary forms, with or without
   * modification, are permitted provided that the following conditions are met:
   *
   * 1. Redistributions of source code must retain the above copyright notice,
   *    this list of conditions and the following disclaimer.
   * 2. Redistributions in binary form must reproduce the above copyright notice,
   *    this list of conditions and the following disclaimer in the documentation
   *    and/or other materials provided with the distribution.
   * 3. Neither the name of STMicroelectronics nor the names of its contributors
   *    may be used to endorse or promote products derived from this software
   *    without specific prior written permission.
   *
   * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
   * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
   * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
   * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
   * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
   * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
   * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
   * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
   * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
   * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
   *******************************************************************************
   */
```

I will copy and paste that on top of every file.

jeromecoutant replied:
Something like this:
````
/* mbed Microcontroller Library
 * SPDX-License-Identifier: BSD-3-Clause
 ******************************************************************************
 *
 * Copyright (c) 2020 STMicroelectronics. 
 * All rights reserved.
 *
 * This software component is licensed by ST under BSD 3-Clause license,
 * the "License"; You may not use this file except in compliance with the 
 * License. You may obtain a copy of the License at:
 *                        opensource.org/licenses/BSD-3-Clause
 *
 ******************************************************************************
 */
````

andreagilardoni replied:
Sorry for that, I do not know much about licensing.
I will try to include this header to every file I added

andreagilardoni committed:
fixing SPDX identifiers in all files
